### PR TITLE
Out-of-memory error now is a termination status.

### PR DESCRIPTION
**Pull request by henriquebecker91**
PR spawned from issue #361. It is basically the same that was done for Gurobi.jl issue [397](https://github.com/jump-dev/Gurobi.jl/issues/397) and PR [398](https://github.com/jump-dev/Gurobi.jl/pull/398). The tests pass:

[](https://postimg.cc/cv2hy9sx)

### CODE REVIEW

**Comment by henriquebecker91** [comment]
Seems like the CPLEX package in CI is not correctly built?

**Comment by dourouc05** [comment]
You need special access rights to run CI, so it's perfectly normal.

**Comment by dourouc05** [comment]
Would you have a very specific case where this error code happens (maybe tunning parameters)? This way, it could be included in the test suite, but it looks like it will be complicated to reproduce in several versions of CPLEX in the CI environment.

**Comment by henriquebecker91** [comment]
hmmm, not sure if I have something that is easy to reproduce. I have a saved MPS model that will trigger it when we try to solve it relaxed (using `JuMP.read_from_file` + `JuMP.relax_integrality` + `JuMP.optimize!`) with the barrier algorithm. But the `mps.gz` has 200MiB (maybe I could reduce it further by making the variable names single-letter).

**Comment by dourouc05** [comment]
If it's so large, I suppose it also takes quite some time before throwing out an error, I don't know if it's worthwhile to add it to the test suite (it would be much slower to run "just" to catch an edge case).

**Comment by henriquebecker91** [comment]
The problem are twofold:

1. If the environment has a lot of memory, the test will take some time, even if it is some pathological case that makes CPLEX memory grow very fast.
2. If the environment has not a lot of memory, then we risk downsizing the test and it will fail in a test environment with more memory.

I think the only clean solution, if testing is deemed necessary, would involve stubbing/mocking the returned value. So we could instantaneously check if the rest of the code works accordingly.

File: src/MOI/MOI_wrapper.jl [hunk 1]
```diff
@@ -2515,6 +2551,9 @@ end
 
 function MOI.get(model::Optimizer, attr::MOI.RawStatusString)
     _throw_if_optimize_in_progress(model, attr)
+    if haskey(_ERROR_TO_STATUS, model.ret_optimize)
+        return _get_error_string(model.env, model.ret_optimize)
```
Comment (odow) [inline src/MOI/MOI_wrapper.jl:2555]: Have you tested this works when called at anytime? I.e., does it depend on some internal state of the `env`?

**Review by odow — APPROVED** [review]
Good by me, modulo the question.

File: src/MOI/MOI_wrapper.jl (hunk 1 above)
Comment (odow) [reply src/MOI/MOI_wrapper.jl:2555]:
```Julia
julia> env = CPLEX.Env()
CPLEX.Env(Ptr{Nothing} @0x00007f9c942a1010, false, 0)

julia> CPLEX._check_ret(env, Cint(CPLEX.CPXERR_NO_MEMORY))
ERROR: CPLEX Error  1001: Out of memory.

Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] _check_ret(::CPLEX.Env, ::Int32) at /Users/oscar/.julia/dev/CPLEX/src/MOI/MOI_wrapper.jl:126
 [3] top-level scope at REPL[8]:1
```

What's weird is
```Julia
julia> CPLEX._check_ret(C_NULL, Cint(CPLEX.CPXERR_NO_MEMORY))
ERROR: CPLEX Error  1001: Out of memory.

Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] _check_ret(::Ptr{Nothing}, ::Int32) at /Users/oscar/.julia/dev/CPLEX/src/MOI/MOI_wrapper.jl:126
 [3] top-level scope at REPL[9]:1
```
Why even have the pointer as the argument?

**Comment by henriquebecker91** [comment]
Added the two tests. The tests pass.

[](https://postimg.cc/QBtDYr60)

**Review by dourouc05 — APPROVED** [review]

**Review by odow — APPROVED** [review]

File: src/MOI/MOI_wrapper.jl (hunk 1 above)
Comment (henriquebecker91) [reply src/MOI/MOI_wrapper.jl:2555]: Well, it is developed by IBM. I would guess that not taking `env` would be considered as exposing the internals, because they would be exposing `env` is not needed. This way if they somehow start needing some info on `env` the call already takes the object.
